heron/instance/src/java/com/twitter/heron/instance/Slave.java

```diff
@@ -183,5 +197,115 @@ public void close() {
     if (instance != null) {
       instance.stop();
     }
+
+    // Clean the resources we own
+    slaveLooper.exitLoop();
+    streamInCommunicator.clear();
+    // The clean of out stream communicator will be handled by instance itself
+  }
+
+  private void handleStartInstanceStatefulProcessing(InstanceControlMsg instanceControlMsg) {
+    CheckpointManager.StartInstanceStatefulProcessing startStatefulRequest =
+        instanceControlMsg.getStartInstanceStatefulProcessing();
+    LOG.info("Starting stateful processing: " + startStatefulRequest.getCheckpointId());
+    isStatefulProcessingStarted = true;
+
+    // At this point, the pre-condition is we have already create the actual instance
+    // Check if we can start the topology
+    startInstanceIfNeeded();
+  }
+
+  private void handleRestoreInstanceStateRequest(InstanceControlMsg instanceControlMsg) {
+    CheckpointManager.RestoreInstanceStateRequest request =
+        instanceControlMsg.getRestoreInstanceStateRequest();
+    LOG.info("Restoring state to checkpoint id: " + request.getState().getCheckpointId());
+    isStatefulProcessingStarted = false;
+    // Reset the in stream queue
+    streamInCommunicator.clear();
+
+    metricsCollector.forceGatherAllMetrics();
+
+    slaveLooper.clearTasksOnWakeup();
+    slaveLooper.clearTimers();
+
+    slaveLooper.addTasksOnWakeup(this);
+    // Create a new MetricsCollector with new clean slave looper
+    metricsCollector = new MetricsCollector(slaveLooper, metricsOutCommunicator);
+
+    // Stop current working instance if there is one
+    if (instance != null) {
+      instance.stop();
+    }
+
+    // Restore the state
+    instanceState.clear();
+    if (request.getState().hasState() && !request.getState().getState().isEmpty()) {
+      // TODO(nlu): Add interface to allow customized serialization
+      @SuppressWarnings("unchecked")
+      Map<byte[], byte[]> stateToRestore = (Map<byte[], byte[]>) Utils.deserialize(
```

Comment:
For what its worth, its better to just piggy back on the regular serialization that ppl use for tuples instead of having multiple serialization interfaces.